### PR TITLE
feat: add preUpgradeTasks configuration option

### PR DESCRIPTION
**Pull request by damien-danglard**
## Changes

This PR change some configuration option :
- add `preUpgradeTasks`
- replace `allowPreUpgradeCommandTemplating` by `allowUpgradeCommandTemplating` (self-hosted)
- replace `allowedPreUpgradeCommands` by `allowedUpgradeCommands` (self-hosted)

## Context

On iOS/macOS development, there is a very important IDE directory named with `.xcodeproj` extension. This directory contains xml files which are regularly source of conflicts. So that's why many iOS/macOS developers use [`xcodegen`](https://github.com/yonaskolb/XcodeGen): a tool that allows us to exclude `*.xcodeproj` directories from git because it can generate it on demand.

Problem: `cocoapods`, our dependency manager requires a `.xcodeproj`; `xcodegen generate` command must be run before `pod install`, but this is currently not possible on Renovate.

Solution: add `preUpgradeTasks` (like `postUpdateTasks`) to allow Renovate to run `xcodegen generate` command before `pod install`

Issues: #8804, #18845, #14179

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by viceice** [comment]
please don't force push, just merge main when required.

please read the contribution guide

File: lib/config/global.ts [hunk 1]
```diff
@@ -5,8 +5,10 @@ export class GlobalConfig {
   private static readonly OPTIONS: (keyof RepoGlobalConfig)[] = [
     'allowCustomCrateRegistries',
     'allowedPostUpgradeCommands',
+    'allowedPreUpgradeCommands',
```
Comment (viceice) [inline lib/config/global.ts:8]: i think we should have one option for this.

eg `allowUpgradeCommands`

@rarkins @secustor @JamieMagee WDYT?

File: lib/config/global.ts [hunk 2]
```diff
@@ -5,8 +5,10 @@ export class GlobalConfig {
...
     'allowPlugins',
     'allowPostUpgradeCommandTemplating',
+    'allowPreUpgradeCommandTemplating',
```
Comment (viceice) [inline lib/config/global.ts:11]: same here

File: lib/config/global.ts (hunk 1 above)
Comment (secustor) [reply lib/config/global.ts:8]: I have tried to come up a scenario in which I would need to split this up but failed at it.   
On the other hand from a UX perspective I think it is better to mirror `pre` and `post` options.

Comment (damien-danglard) [reply lib/config/global.ts:8]: If I remove the `allowedPostUpgradeCommands` and `allowedPreUpgradeCommands` and add `allowedPeriUpgradeCommands` (for example), won't this cause backward compatibility issues (for those who already use `allowedPostUpgradeCommands`)?

Moreover, the commands executed in pre-update and in post-update have no interest in being the same. Post-update commands are mainly controls (ex: lint) while pre-update commands are mainly content generation commands (ex: xcodegen generates a directory required by cocoapods).

Comment (viceice) [reply lib/config/global.ts:8]: i only mean the allow options, not the actual repo commands. we can migrate existing config options automatically.

so if we allow pre commands, i don't see a use case to not allow post commands, as both can nearly do anything.

Comment (rarkins) [reply lib/config/global.ts:8]: I think it would be best to have a unified "allowed commands" concept which applies to both pre and post

Comment (damien-danglard) [reply lib/config/global.ts]: What about my last commit: [refactor: unify allowPostUpgradeCommandTemplating/allowPreUpgradeComm…](https://github.com/renovatebot/renovate/pull/20423/commits/760a6999df54b9cd9cfee4727ad9c3ddb205e5ee) ?

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -2181,7 +2181,7 @@ Table with options:
 Post-upgrade tasks are commands that are executed by Renovate after a dependency has been updated but before the commit is created.
 The intention is to run any additional command line tools that would modify existing files or generate new files when a dependency changes.
 
-Each command must match at least one of the patterns defined in `allowedPostUpgradeCommands` (a global-only configuration option) in order to be executed.
+Each command must match at least one of the patterns defined in `allowedPeriUpgradeCommands` (a global-only configuration option) in order to be executed.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2184]: | Term   | Meaning |
| ------ | --------|
| `peri` | Around  |
| `pre`  | Before  |

> `allowedPeriUpgradeCommands`

I think you meant allowed**Pre**UpgradeCommands? 😉 

There are more places where you use `peri`, so search and replace them if needed.

File: lib/config/global.ts [hunk 4]
```diff
@@ -4,9 +4,9 @@ export class GlobalConfig {
   // TODO: once global config work is complete, add a test to make sure this list includes all options with globalOnly=true (#9603)
   private static readonly OPTIONS: (keyof RepoGlobalConfig)[] = [
     'allowCustomCrateRegistries',
-    'allowedPostUpgradeCommands',
+    'allowedPeriUpgradeCommands',
```
Comment (viceice) [inline lib/config/global.ts:7]:
```suggestion
    'allowedUpgradeCommands',
```

File: lib/config/global.ts [hunk 5]
```diff
@@ -4,9 +4,9 @@ export class GlobalConfig {
...
     'allowPlugins',
-    'allowPostUpgradeCommandTemplating',
+    'allowPeriUpgradeCommandTemplating',
```
Comment (viceice) [inline lib/config/global.ts:9]:
```suggestion
    'allowUpgradeCommandTemplating',
```

**Review by viceice — COMMENTED** [review]
🤔

File: docs/usage/configuration-options.md (hunk 3 above)
Comment (damien-danglard) [reply docs/usage/configuration-options.md:2184]: No, because @rarkins said he prefers using a single config argument for commands allowed for pre-upgrade and post-upgrade tasks ([see conversation](https://github.com/renovatebot/renovate/pull/20423#discussion_r1108093615)).

So that's why I changed all `allowedPreUpgradeCommands` and `allowedPostUpgradeCommands` to `allowedPeriUpgradeCommands`. `peri` means `pre` and `post` in this context.

File: lib/config/global.ts (hunk 4 above)
Comment (damien-danglard) [reply lib/config/global.ts:7]: I used the peri prefix so that there is no ambiguity. I wouldn't want users to think that it does not include upgrade commands allowed in the "allowedUpgradeCommands" property, such as `"allowedUpgradeCommands": ["pod install", "npm install", "mvn install"]`.

If you think there will be no risk of ambiguity, that's fine for me.

File: docs/usage/configuration-options.md (hunk 3 above)
Comment (HonkingGoose) [reply docs/usage/configuration-options.md:2184]: If we want a single command for both `pre` and `post` phases then `peri` is good. 😉

Comment (rarkins) [reply docs/usage/configuration-options.md]: I think it's too obscure, we'd be better with `allowedUpgradeCommands`

File: lib/config/global.ts [hunk 6]
```diff
@@ -4,9 +4,9 @@ export class GlobalConfig {
   // TODO: once global config work is complete, add a test to make sure this list includes all options with globalOnly=true (#9603)
   private static readonly OPTIONS: (keyof RepoGlobalConfig)[] = [
     'allowCustomCrateRegistries',
-    'allowedPostUpgradeCommands',
+    'allowedUpgradeCommands',
     'allowPlugins',
-    'allowPostUpgradeCommandTemplating',
+    'allowUpgradeCommandTemplating',
```
Comment (viceice) [inline lib/config/global.ts:9]: please sort list

File: docs/usage/configuration-options.md [hunk 7]
```diff
@@ -2200,22 +2200,64 @@ e.g.
 
 The `postUpgradeTasks` configuration consists of three fields:
 
-### commands
+### commands (postUpgradeTasks)
```
Comment (viceice) [inline docs/usage/configuration-options.md:2203]: I'm not sure if the docs generation works when title doesn't equal the config option.

@rarkins maybe i need to update it before to allow title suffix

File: lib/config/options/index.ts [hunk 8]
```diff
@@ -79,6 +79,47 @@ const options: RenovateOptions[] = [
...
+      executionMode: 'update',
+    },
+  },
+  {
+    name: 'commands',
+    description:
+      'A list of pre-upgrade commands that are executed before updating dependency files by Renovate.',
+    type: 'array',
+    subType: 'string',
+    parent: 'preUpgradeTasks',
```
Comment (viceice) [inline lib/config/options/index.ts:99]: @rarkins I think we need to allow multiple parents, so we need to define commands and others only once.

File: lib/config/types.ts [hunk 9]
```diff
@@ -153,7 +154,7 @@ export interface LegacyAdminConfig {
 
 export type ExecutionMode = 'branch' | 'update';
 
-export interface PostUpgradeTasks {
+export interface PeriUpgradeTasks {
```
Comment (viceice) [inline lib/config/types.ts:157]:
```suggestion
export interface UpgradeTasks {
```

File: lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+import { sanitize } from '../../../../util/sanitize';
+import { compile } from '../../../../util/template';
+import type { BranchConfig, BranchUpgradeConfig } from '../../../types';
+
+export interface PreUpgradeCommandsExecutionResult {
+  updatedArtifacts: FileChange[];
+  artifactErrors: ArtifactError[];
+}
+
+export async function preUpgradeCommandsExecutor(
```
Comment (viceice) [inline lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts:25]: refactor post upgrade tasks to be more generic, so we don't need to duplicate all that code for pre upgrade tasks.

refactor should probably be done in a separate PR.

File: lib/config/options/index.ts [hunk 11]
```diff
@@ -22,17 +22,17 @@ const options: RenovateOptions[] = [
     globalOnly: true,
   },
   {
-    name: 'allowPostUpgradeCommandTemplating',
+    name: 'allowUpgradeCommandTemplating',
```
Comment (viceice) [inline lib/config/options/index.ts:25]: needs migration. simply add them to this list:

https://github.com/renovatebot/renovate/blob/59432b412973a86d99ce7319d93d557473b781db/lib/config/migrations/migrations-service.ts#L75

File: lib/config/options/index.ts [hunk 12]
```diff
@@ -22,17 +22,17 @@ const options: RenovateOptions[] = [
...
     description:
-      'Set this to `true` to allow templating for post-upgrade commands.',
+      'Set this to `true` to allow templating for pre-upgrade and post-upgrade commands.',
     type: 'boolean',
     default: false,
     globalOnly: true,
   },
   {
-    name: 'allowedPostUpgradeCommands',
+    name: 'allowedUpgradeCommands',
```
Comment (viceice) [inline lib/config/options/index.ts:33]: needs migration.

**Review by viceice — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md (hunk 7 above)
Comment (damien-danglard) [reply docs/usage/configuration-options.md:2203]: I think it's a markdown issue if they are multiple identical title (`commands` from `preUpgradeTasks` and from `postUpgradeTasks`) so it's for this a have added suffix.

File: docs/usage/configuration-options.md [hunk 13]
```diff
@@ -2200,22 +2200,64 @@ e.g.
...
+### fileFilters (preUpgradeTasks)
+
+A list of glob-style matchers that determine which files will be included in the final commit made by Renovate.
+
+### executionMode (preUpgradeTasks)
+
+Defaults to `update`, but can also be set to `branch`.
+This sets the level the preUpgradeTask runs on, if set to `update` the preUpgradeTask will be executed for every dependency on the branch.
+If set to `branch` the preUpgradeTask is executed for the whole branch.
+
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2260]: If `preUpgradeTasks` is for self-hosted only, then move the section into `docs/usage/self-hosted-configuration.md`. 😉

File: docs/usage/configuration-options.md [hunk 14]
```diff
@@ -2200,22 +2200,64 @@ e.g.
...
 
 Defaults to `update`, but can also be set to `branch`.
 This sets the level the postUpgradeTask runs on, if set to `update` the postUpgradeTask will be executed for every dependency on the branch.
 If set to `branch` the postUpgradeTask is executed for the whole branch.
 
+## preUpgradeTasks
+
+<!-- prettier-ignore -->
+!!! note
+    Pre-upgrade tasks can only be used on self-hosted Renovate instances.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2223]: Once you move the whole section to the self-hosted config option docs, you can drop this admonition. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md (hunk 13 above)
Comment (HonkingGoose) [reply docs/usage/configuration-options.md:2260]: If `postUpgradeTasks` is meant for self-hosted-only, then also move that section to `docs/usage/self-hosted-configuration.md`.

Comment (damien-danglard) [reply docs/usage/configuration-options.md:2260]: Yes it is (https://github.com/renovatebot/renovate/pull/20423/files#diff-be0a39f342380493f17c380bbd694f1f05d18bdd8d7cecf174369084952102deL2177-L2181).

File: lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts (hunk 10 above)
Comment (damien-danglard) [reply lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts:25]: I have an another branch with refactoring (https://github.com/damien-danglard/renovate/pull/2). I can merge it in this one or make an another PR, as you want.

I think it's better to make an another PR because it's a big refactoring.

File: docs/usage/self-hosted-configuration.md [hunk 15]
```diff
@@ -555,13 +555,89 @@ It also may mean that ignored directories like `node_modules` can be preserved a
 
 ## platform
 
+## postUpgradeTasks
```
Comment (viceice) [inline docs/usage/self-hosted-configuration.md:558]: no, this needs to be moved back to other config docs, as it's a repo config option.

only options with global flag should be defined here.

File: lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts (hunk 10 above)
Comment (viceice) [reply lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts:25]: yes, create a separate PR for refactoring first.
so we don't have to duplicate so much code

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
converted to draft, as we need a refactor PR first.

File: docs/usage/self-hosted-configuration.md (hunk 15 above)
Comment (HonkingGoose) [reply docs/usage/self-hosted-configuration.md:558]: Sorry about giving the wrong instructions! 🙈

File: lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts (hunk 10 above)
Comment (damien-danglard) [reply lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts:25]: I do this : https://github.com/renovatebot/renovate/pull/20571
@viceice  can you review it please ?

File: docs/usage/self-hosted-configuration.md (hunk 15 above)
Comment (damien-danglard) [reply docs/usage/self-hosted-configuration.md:558]: @HonkingGoose no problem, I just need to revert commit :)

File: docs/usage/configuration-options.md (hunk 7 above)
Comment (HonkingGoose) [reply docs/usage/configuration-options.md:2238]: I think the headings must match the config option name as defined in the source code. I recently got shouted at by the docs tests for doing it wrong. 😄

File: lib/config/options/index.ts [hunk 16]
```diff
@@ -79,6 +79,47 @@ const options: RenovateOptions[] = [
...
+    type: 'array',
+    subType: 'string',
+    parent: 'preUpgradeTasks',
+    default: [],
+    cli: false,
+  },
+  {
+    name: 'executionMode',
+    description:
+      'Controls when the pre upgrade tasks run: on every update, or once per upgrade branch.',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:116]:
```suggestion
      'Controls when the pre-upgrade tasks run: on every update, or once per upgrade branch.',
```

File: docs/usage/configuration-options.md [hunk 17]
```diff
@@ -2235,22 +2235,64 @@ e.g.
...
+
+```json
+{
+  "preUpgradeTasks": {
+    "commands": ["xcodegen generate"],
+    "fileFilters": [],
+    "executionMode": "update"
+  }
+}
+```
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2276]: It's not clear to me if this is an example of the _good_ way, or the _bad_ way to configure the bot.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some more comments to move this a bit further, but still needs a more thorough review after the refactoring PR landed. I doubt that all the text is in the correct place for example.

**Comment by rarkins** [comment]
Closing due to inactivity and conflicts
